src/libsyntax/ext/expand.rs

```diff
@@ -531,11 +532,35 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         let path = &mac.node.path;
 
         let ident = ident.unwrap_or_else(|| keywords::Invalid.ident());
-        let validate_and_set_expn_info = |def_site_span,
+        let validate_and_set_expn_info = |self_: &mut Self, // arg instead of capture
+                                          def_site_span,
                                           allow_internal_unstable,
-                                          allow_internal_unsafe| {
+                                          allow_internal_unsafe,
+                                          // can't infer this type
+                                          unstable_feature: Option<(Symbol, u32)>| {
+
+            // feature-gate the macro invocation
+            if let Some((feature, issue)) = unstable_feature {
+                let crate_span = self_.cx.current_expansion.crate_span.unwrap();
+                // don't stability-check macros in the same crate
+                if !crate_span.contains(def_site_span)
```

Comment:
I don't know how this is going to behave in corner cases (e.g. macros defined by macros), but this seems to be the conservative solution (if one span contains another, they are in one crate) and I'll r+ this if nobody comes up with a better solution until in few days.

Reply:
I was going to add the macro-generating-macro case to the test as well.